nbs/03_matcher.ipynb

```diff
@@ -165,6 +165,9 @@
    "outputs": [],
```

Comment:
<pre class="ql-syntax" spellcheck="false">def find_structural_matches(pattern,g):

  possible_candidates={
  node_name:nodes_that_match_on_attributes(pattern[node_name],g)
  }

  for node_options in itertools.product(node,possible_candidates.values()):
  subgraph = g.subgraph(node_options)
  if isomorphism (pattern,sub_graph)
    yield subgraph

example pattern:
- { 'x':{a=b}, 'y':{}} # first phase
- { 'x':{a=b,_id=1}, 'y':{}} # second phase
 
</pre><p> </p>

<pre class="ql-syntax" spellcheck="false">def find_match(pattern,collection_pattern,g,condition_funcs):
  
   for phase_1_iso in  find_structural_matches(pattern,g):
      phase_1_match = pattern.copy()
      # add _id constrains from phase_1_iso
         # find intersection nodes ....
      phase_2_isos = [phase_2_iso in find_structural_matches(collection_pattern_with_id_constraints,g)]  
      match = Match(regular_nodes = pahse_1_iso,collections = phase_2_isos)  
      if condition(match):
         yield match
</pre><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/DeanLight/graph_rewrite/pull/17/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>